google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/BigtableUnaryOperationCallable.java

```diff
@@ -0,0 +1,189 @@
+    private final ApiTracer tracer;
+    private final boolean allowNoResponse;
+
+    private StreamController controller;
+    private final AtomicBoolean upstreamCancelled = new AtomicBoolean();
+    private boolean responseReceived;
+    private @Nullable RespT response;
+
+    private UnaryFuture(ApiTracer tracer, boolean allowNoResponse) {
+      this.tracer = Preconditions.checkNotNull(tracer, "tracer can't be null");
+      this.allowNoResponse = allowNoResponse;
+      this.responseReceived = false;
+    }
+
+    @Override
+    public void onStart(StreamController controller) {
+      this.controller = controller;
+      controller.disableAutoInboundFlowControl();
+      // Request 2 to detect protocol bugs
+      controller.request(2);
+    }
+
+    /**
+     * Immediately cancel the future state and try to cancel the underlying operation. Will return
+     * false if the future is already resolved.
+     */
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning) {
+      if (super.cancel(mayInterruptIfRunning)) {
+        cancelUpstream();
+        return true;
+      }
+      return false;
+    }
+
+    private void cancelUpstream() {
+      if (upstreamCancelled.compareAndSet(false, true)) {
+        controller.cancel();
+      }
+    }
+
+    @Override
+    public void onResponse(RespT resp) {
+      tracer.responseReceived();
+
+      // happy path - buffer the only responsse
+      if (!responseReceived) {
+        responseReceived = true;
+        this.response = resp;
+        return;
+      }
+
+      String msg =
+          String.format(
+              "Received multiple responses for a %s unary operation. Previous: %s, New: %s",
+              spanName, response, resp);
+      logger.log(Level.WARNING, msg);
+
+      InternalException error =
+          new InternalException(msg, null, GrpcStatusCode.of(Status.Code.INTERNAL), false);
+      if (setException(error)) {
+        tracer.operationFailed(error);
+      }
+
+      cancelUpstream();
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+      if (this.setException(throwable)) {
+        tracer.operationFailed(throwable);
+      } else if (isCancelled()) {
+        tracer.operationCancelled();
+      }
+      // The future might've been resolved due to double response
+    }
+
+    @Override
+    public void onComplete() {
+      if (allowNoResponse || responseReceived) {
+        if (set(response)) {
+          tracer.operationSucceeded();
+          return;
+        }
+      } else {
+        String msg = spanName + " unary operation completed without a response message";
+        InternalException e =
+            new InternalException(msg, null, GrpcStatusCode.of(Status.Code.INTERNAL), false);
+
+        if (setException(e)) {
+          tracer.operationFailed(e);
+          return;
+        }
+      }
+
+      // check cancellation race
+      if (isCancelled()) {
+        tracer.operationCancelled();
+      }
```

Comment:
Should we also do tracer.operationFailed here?  if all of (set(Response)) or setException(e) and isCancelled returned false we'll end up not calling operation completion in the tracer? I'm not sure when we'll hit this case though. Same question for onError. Do we trust that onComplete or onError is already called by another thread?

Reply:
no, because that will double call the tracer

if all of those methods returned false it means we've already notified the tracer.  

At the beginning of the method we can be in one of 2 states:
1. unresolved (happy path)
2. error (onResponse was called multiple times)
3. cancelled (user cancelled the future)

case#1 is straightfoward and is handled by the set on L168
case#2 means that onResponse setException() and paired it with a tracer notification, so we dont want to notify the tracer again

case#3 would make set & stException return false and we explicitly detect it on 184